src/evict/evict_page.c

```diff
@@ -748,28 +763,37 @@ __evict_review_obsolete_time_window(WT_SESSION_IMPL *session, WT_REF *ref)
     else if (__wt_ref_addr_copy(session, ref, &addr))
         WT_TIME_AGGREGATE_COPY(&newest_ta, &addr.ta);
 
-    /*
-     * Mark the page as dirty to allow the page reconciliation to remove all information related to
-     * an obsolete time window if the page has not been fully removed. The pages that are totally
-     * removed are eliminated during the checkpoint cleanup procedure.
-     */
-    if (!WT_TIME_AGGREGATE_HAS_STOP(&newest_ta) && newest_ta.newest_txn != WT_TXN_NONE &&
-      __wt_txn_visible_all(session, newest_ta.newest_txn,
-        WT_MAX(newest_ta.newest_start_durable_ts, newest_ta.newest_stop_durable_ts))) {
-        __wt_verbose(session, WT_VERB_EVICT,
-          "%p in-memory page obsolete time window: time aggregate %s", (void *)ref,
-          __wt_time_aggregate_to_string(&newest_ta, time_string));
+    /* Pages that are totally removed are eliminated during the checkpoint cleanup procedure. */
+    if (WT_TIME_AGGREGATE_HAS_STOP(&newest_ta))
+        return (0);
 
-        WT_RET(__wt_page_modify_init(session, ref->page));
-        __wt_page_modify_set(session, ref->page);
+    /* If there is no transaction or timestamp information available, there is nothing to do. */
+    max_durable_ts = WT_MAX(newest_ta.newest_start_durable_ts, newest_ta.newest_stop_durable_ts);
+    if (newest_ta.newest_txn == WT_TXN_NONE && max_durable_ts == WT_TS_NONE)
+        return (0);
 
-        /*
-         * To prevent the race while incrementing this variable, no atomic functions are required.
-         * More clean pages may become dirty as a result of an outdated value.
-         */
-        S2BT(session)->obsolete_tw_pages++;
-        WT_STAT_CONN_DSRC_INCR(session, cache_eviction_dirty_obsolete_tw);
-    }
+    /* Ensure the time window information has content that is globally visible. */
+    if (!__wt_txn_visible_all(session, newest_ta.newest_txn, max_durable_ts))
+        return (0);
+
+    /* Mark the page dirty to remove any obsolete information during reconciliation. */
+    __wt_verbose(session, WT_VERB_EVICT,
+      "%p in-memory page obsolete time window: time aggregate %s", (void *)ref,
+      __wt_time_aggregate_to_string(&newest_ta, time_string));
+
+    WT_RET(__wt_page_modify_init(session, ref->page));
+    __wt_page_modify_set(session, ref->page);
+
+    /* Save that a new btree has been processed. */
+    if (btree->obsolete_tw_pages == 0)
```

Comment:
These variables are shared. There is a potential race here for these variables. I don't know how accurate these counters need to be but at least we should mark them as shared and use atomic operations to make TSAN happy.

Reply:
I do not think that we need these values to be accurate. We are not going to cause any problem if we do 1 or 2 btrees more/less dirty than the configured ones. Adding atomics/barriers are unnecessarily going to affect the performance.

Reply:
That was my understanding too - should we add a comment somewhere to make it clear?

Reply:
The value don't need to be accurate. But TSAN build will fail with the current code. We should use relaxed atomic operations for this. The relaxed operations will not cause any performance impact. It is only to keep TSAN happy.

Reply:
I tried a TSAN [patch build](https://spruce.mongodb.com/task/wiredtiger_ubuntu2004_tsan_examples_c_tsan_patch_f0c94a7fc54a4a22d26cf85668917020a9459a85_668fce6f5f4d2400074cff65_24_07_11_12_22_21/logs?execution=0) but it does not fail on this code.
@ajmorton, will TSAN fail if I commit this code? The race is quite obvious but how do we flag it as "it's ok"?

Reply:
A single TSAN run will not fail. This will only fail when TSAN actually hit this race.

Reply:
I would believe it to be super likely but fair enough, I can try multiple runs.

Reply:
I would believe it to be unlikely.

Reply:
Yep TSan will fail if we merge the current code.

Here's a reduced extract from the logs where I removed most of the call stack
```
WARNING: ThreadSanitizer: data race (pid=10398)
Read of size 4 at 0x7b9400000bf8 by thread T5 (mutexes: write M35):
    #0 __evict_review_obsolete_time_window /data/mci/963b7dbeeef58d7a062f55e836667798/wiredtiger/src/evict/evict_page.c:741:32 (libwiredtiger.so.11.3.0+0x207eca)
    #1 __evict_review /data/mci/963b7dbeeef58d7a062f55e836667798/wiredtiger/src/evict/evict_page.c:838:5 
  
Previous write of size 4 at 0x7b9400000bf8 by main thread (mutexes: write M338, write M3):
    #0 __txn_checkpoint /data/mci/963b7dbeeef58d7a062f55e836667798/wiredtiger/src/txn/txn_ckpt.c:1162:54 (libwiredtiger.so.11.3.0+0x344735)
    #1 __txn_checkpoint_wrapper /data/mci/963b7dbeeef58d7a062f55e836667798/wiredtiger/src/txn/txn_ckpt.c:1568:11 (libwiredtiger.so.11.3.0+0x3429e1)
```

The important lines are the ones starting with `#0`. If you look at them in the code base they are:
`      conn->heuristic_controls.obsolete_tw_btree_count >=`
and
`    conn->heuristic_controls.obsolete_tw_btree_count = 0;`
so we know there's a data race on `obsolete_tw_btree_count`.

There's no good way to tell TSan "ignore this specific struct field" since data races are undefined behaviour. 
The two options to address this are either suppressing the entire `__evict_review_obsolete_time_window` function - which I strongly recommend against - or using our `__wt_atomic_*` functions whenever we touch `obsolete_tw_btree_count`. Skipping over some details we haven't seen performance impacts of using `__wt_atomic_*` functions over normal reads and writes so you don't need to worry about perf here.

Reply:
Thank you, @ajmorton! I will update the functions related to `heuristic_controls` so satisfy TSAN.

Reply:
I have made the updates in 49e3b592778854abd5b56efa4bd0d1f069e8c69e, please let me know if I forgot anything @ajmorton.